lib/scanner/walk.go

```diff
@@ -98,30 +111,61 @@ type walker struct {
 
 // Walk returns the list of files found in the local folder by scanning the
 // file system. Files are blockwise hashed.
-func (w *walker) walk(ctx context.Context) (chan protocol.FileInfo, error) {
+func (w *walker) walk(ctx context.Context) (chan ScanResult, error) {
 	l.Debugln("Walk", w.Subs, w.BlockSize, w.Matcher)
 
 	if err := w.checkDir(); err != nil {
 		return nil, err
 	}
 
-	toHashChan := make(chan protocol.FileInfo)
-	finishedChan := make(chan protocol.FileInfo)
+	fsChan := make(chan fsWalkResult)
+
+	haveChan := make(chan *protocol.FileInfo)
+	haveCtx, haveCancel := context.WithCancel(ctx)
+
+	// A routine which walks the db and returns file infos to be compared
+	// to scan results.
+	go func() {
```

Comment:
I think this, and its peer below, would do nicely as actual named methods?

(And we can probably defer the channel close / context cancel in those functions.)